x-pack/plugins/security_solution/docs/prebuilt_rules_customization_rfc.md

```diff
@@ -0,0 +1,2867 @@
+    ...typeSpecificCamelToSnake(rule.params),
+    // [... more object properties ...]
+  };
+};
+
+export const commonParamsCamelToSnake = (params: BaseRuleParams) => {
+  const { immutable, rule_source } = normalizeRuleSourceSchemaOnRuleRead(params);
+
+  return {
+    immutable,
+    rule_source,
+    // [... more object properties ...]
+  };
+};
+```
+
+And the `normalizeRuleSourceSchemaOnRuleRead` can be defined so:
+
+_Source: x-pack/plugins/security_solution/server/lib/detection_engine/rule_management/normalization/prebuilt_rule_schema_migrations.ts_ (New file)
+
+```ts
+interface OnReadNormalizationResult {
+  immutable: IsRuleImmutable;
+  rule_source: RuleSource
+}
+
+/**
+ * To calculate `rule_source`:
+ * - Use `rule_source` field  if already exists in the internal rule object params.
+ * - If it does not exist, check the value `immutable` from the rules param:
+ *   - if it is `true`, create a new `rule_source` object, with `isCustomized` set to `false` and `repoNam` set to `elastic_prebuilt` (Use case of rule_source rules that have not been yet migrated-on-write)
+ *   - otherwise, the field should be `undefined` (use case of custom rules).
+ */
+const getRuleSourceValueForRuleRead = (ruleParams: BaseRuleParams): RuleSource | undefined => {
+  if (ruleParams.rule_source) {
+    return ruleParams.rule_source;
+  }
+
+  if (ruleParams.immutable) {
+    return {
+      type: 'external'
+      isCustomized: false,
+    };
+  }
+
+  return {
+    type: 'internal'
+  };
+};
+
+export const normalizePrebuiltSchemaOnRuleRead = (
+  ruleParams: BaseRuleParams
+): OnReadNormalizationResult => {
+
+/**
+ * To calculate `immutable`:
+ * - Checks if the `rule_source` field exists in the rule's parameters.
+ *   - If `rule_source` exists and its type is external, sets `immutable` to `true`. (Use case of Rules that have already been migrated-on-write)
+ *   - If `rule_source` does not exist, return the value of the params' `immutable` field. (Use case of Rules that have not yet been migrated on write.)
+ */
+  const isRuleSourceExternal = ruleParams.rule_source && ruleParams.rule_source.type === 'external';
+  const immutable = Boolean(isRuleSourceExternal) || Boolean(ruleParams.immutable);
+  const rule_source = getRuleSourceValueForRuleRead(ruleParams);
+
+  return {
+    immutable,
+    rule_source,
+  };
+};
+```
+
+##### Rule Management endpoints that will perform normalization-on-read
+
+- All endpoints that also perform migration-on-write, as listed in the below `Migration on write` section. All of them use the `internalRuleToAPIResponse` methods to transform the rule before returning the response, so the normalization on read will take place there. For these specific cases, however, the normalization-on-read will take place, but the data stored in ES for each rule should match the response given by the endpoint, i.e. the normalized in-memory value and the stored value of the rule should be the same.
```

Comment:
The text here looks overcomplicated. I'd recommend to split this section in two sub parts read API endpoints and delete/bulk endpoints. The latter return rule's data and need to make sure returned data is normalized.

Btw, what's the problem to perform migration-on-write when enabling/disabling rules? Could you elabored on the reasons in the RFC?

Reply:
Bulk enabling and disabling rules use the RuleClient's `enable` and `disable`  methods. We pass in only the `id` and it fetches the attributes and does the modification of the `enabled` prop.

Doing migration-on-write is techninally possible, but would mean adding the migration-on-write logic on the RuleClient side, which I would prefer to avoid. I'd like to keep all migration logic on the Security solution side. 

On the other side, enabling and disabling are probably the most-often executed actions, so if we want to accelerate the migration of a user's SO, maybe we should actually do this.

What do you think? cc @banderror

Reply:
> Doing migration-on-write is techninally possible, but would mean adding the migration-on-write logic on the RuleClient side, which I would prefer to avoid.

@jpdjere I also think we should avoid this. The RulesClient doesn't and should not know about our rule params. Properly implementing rule params migrations on enable/disable would require creating a special mechanism for that, which doesn't seem to worth it.

Besides that, we're going to migrate to the bulk enable/disable methods of the RulesClient very soon: https://github.com/elastic/kibana/issues/177634.

Reply:
- Detail in RFC why not doing migration-on-write on bulk enabling/disabling.
- Create a separate ticket for evaluating migration-on-write on enabling/disabling for small amounts.